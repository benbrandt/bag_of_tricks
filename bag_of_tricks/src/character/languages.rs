use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use super::Character;

/// Available languages for a character to learn.
#[allow(dead_code)]
#[derive(Clone, Copy, Deserialize, Display, EnumIter, Eq, Hash, PartialEq, Serialize)]
pub(crate) enum Language {
    Abyssal,
    Auran,
    Celestial,
    Common,
    #[strum(serialize = "Deep Speech")]
    DeepSpeech,
    Draconic,
    Dwarvish,
    Elvish,
    Giant,
    Gnomish,
    Goblin,
    Halfling,
    Infernal,
    Orc,
    Primordial,
    Sylvan,
    Undercommon,
}

impl Language {
    // Sometimes you end up with dupes. Consume and replace with a new option
    pub(crate) fn gen(rng: &mut impl Rng, character: &Character, amount: usize) -> Vec<Self> {
        Language::iter()
            .filter(|l| !character.languages.contains(l))
            .choose_multiple(rng, amount)
    }
}

/// Trait for describing what languages are provided by a given object.
pub(crate) trait Languages {
    /// Return list of languages for this object
    fn languages(&self) -> Vec<Language> {
        vec![]
    }

    /// Additional languages this character can learn.
    fn addl_languages(&self) -> usize {
        0
    }
}
