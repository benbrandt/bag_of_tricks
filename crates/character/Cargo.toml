[package]
name = "character"
version = "0.1.0"
edition = "2018"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
alignment = { path = "../alignment" }
attack = { path = "../attack" }
characteristics = { path = "../characteristics" }
citation = { path = "../citation" }
deities = { path = "../deities" }
dice_roller = { path = "../dice_roller" }
itertools = "0.10"
rand = "0.8"
rand_pcg = "0.3"
serde = { version = "1", features = ["derive"] }
strum = { version = "0.20", features = ["derive"] }
typetag = "0.1"

[dev-dependencies]
insta = "1"
serde_yaml = "0.8"
