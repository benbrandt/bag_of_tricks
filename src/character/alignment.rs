use std::fmt;

use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Attitude {
    Chaotic,
    Lawful,
    Neutral,
}

#[derive(Deserialize, Display, EnumIter, Serialize)]
enum Morality {
    Evil,
    Good,
    Neutral,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Alignment(Attitude, Morality);

impl Alignment {
    pub(crate) fn gen(rng: &mut impl Rng) -> Self {
        let attitude = Attitude::iter().choose(rng).unwrap();
        let morality = Morality::iter().choose(rng).unwrap();
        Self(attitude, morality)
    }
}

impl fmt::Display for Alignment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self(Attitude::Neutral, Morality::Neutral) => write!(f, "Neutral"),
            Self(a, m) => write!(f, "{} {}", a, m),
        }
    }
}
