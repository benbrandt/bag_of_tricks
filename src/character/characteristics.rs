use rand::{prelude::IteratorRandom, Rng};
use serde::{Deserialize, Serialize};
use std::{fmt, ops::RangeInclusive};
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use crate::dice_roller::RollCmd;

/// An range of ages a given adventurer could be.
pub(crate) struct AgeRange(pub(crate) RangeInclusive<u16>);

impl AgeRange {
    /// Generate a random age within the given range.
    fn gen(&self, rng: &mut impl Rng) -> u16 {
        rng.gen_range(self.0.clone())
    }
}

/// Really only here to help decide on names, not core to a character choice.
#[derive(Deserialize, Display, EnumIter, Serialize)]
pub(crate) enum Gender {
    Female,
    Male,
}

impl Gender {
    /// Choose a random gender
    fn gen(rng: &mut impl Rng) -> Self {
        Gender::iter().choose(rng).unwrap()
    }
}

/// Weight modifier
#[derive(Debug, PartialEq)]
pub(crate) enum WeightMod {
    /// Fixed additional weight
    Fixed(u16),
    /// Role to find additional weight
    Roll(RollCmd),
}

/// Convert feet + inches to just inches (allow for better readability in tables)
pub(crate) const fn in_inches(feet: u8, inches: u8) -> u8 {
    feet * 12 + inches
}

/// Data from a table to generate a height and weight for a given character.
#[derive(Debug, PartialEq)]
pub(crate) struct HeightAndWeightTable {
    /// Base height for this character
    pub(crate) base_height: u8,
    /// Base weight for this character
    pub(crate) base_weight: u16,
    /// Modifier to apply to the height
    pub(crate) height_mod: RollCmd,
    /// Modifier to apply to the weight (informed by height)
    pub(crate) weight_mod: WeightMod,
}

impl HeightAndWeightTable {
    /// Generate a random height and weight for a character
    pub(crate) fn gen(&self, rng: &mut impl Rng) -> (usize, usize) {
        let h = self.height_mod.roll(rng).total();
        // Weight modifier is multiplied by height
        let w = h * match &self.weight_mod {
            WeightMod::Fixed(f) => usize::from(*f),
            WeightMod::Roll(r) => r.roll(rng).total(),
        };
        (
            usize::from(self.base_height) + h,
            usize::from(self.base_weight) + w,
        )
    }
}

/// Size of character (there are more options for monsters)
#[derive(Deserialize, Display, EnumIter, Serialize)]
pub(crate) enum Size {
    Small,
    Medium,
}

/// Types of movement speeds
#[derive(Clone, Copy, Debug, Deserialize, EnumIter, PartialEq, Serialize)]
pub(crate) enum Speed {
    Climbing(u8),
    Swimming(u8),
    Walking(u8),
}

impl fmt::Display for Speed {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::Climbing(s) => write!(f, "Climbing Speed: {}ft", s),
            Self::Swimming(s) => write!(f, "Swimming Speed: {}ft", s),
            Self::Walking(s) => write!(f, "Walking Speed: {}ft", s),
        }
    }
}

/// Physical characteristics about a character.
#[derive(Deserialize, Serialize)]
pub(crate) struct CharacteristicDetails {
    /// Age of the character
    pub(crate) age: u16,
    /// Base speed of the character
    pub(crate) base_speeds: Vec<Speed>,
    /// Gender of the character (only used for name choices)
    pub(crate) gender: Gender,
    /// Height of the character
    pub(crate) height: usize,
    /// Size of the character
    pub(crate) size: Size,
    /// Weight of the character
    pub(crate) weight: usize,
}

impl fmt::Display for CharacteristicDetails {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "Age: {}", self.age)?;
        writeln!(f, "Gender: {}", self.gender)?;
        writeln!(f, "Size: {}", self.size)?;
        writeln!(f, "Height: {}'{}\"", self.height / 12, self.height % 12)?;
        writeln!(f, "Weight: {} lb.", self.weight)
    }
}

/// Trait for object (usually race) to generate appropriate characteristics about a character
///
/// Separate from race since it uses associated constants (which can't be on a trait object)
pub(crate) trait Characteristics {
    /// Age range to choose from
    const AGE_RANGE: AgeRange;
    /// Character size
    const SIZE: Size;

    /// Base speeds
    fn get_base_speeds(&self) -> Vec<Speed>;

    /// Calculate the height and weight table to generate from
    fn get_height_and_weight_table(&self) -> &HeightAndWeightTable;

    /// Generate characteristics for this race
    fn gen_characteristics(&self, rng: &mut impl Rng) -> CharacteristicDetails {
        let (height, weight) = self.get_height_and_weight_table().gen(rng);
        CharacteristicDetails {
            age: Self::AGE_RANGE.gen(rng),
            base_speeds: self.get_base_speeds(),
            gender: Gender::gen(rng),
            height,
            size: Self::SIZE,
            weight,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_age_range() {
        let mut rng = Pcg64::from_entropy();
        let age = AgeRange(1..=100).gen(&mut rng);
        assert!((1..=100).contains(&age));
    }

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_characteristics() {
        let mut rng = Pcg64::seed_from_u64(1);
        let characteristics = CharacteristicDetails {
            age: 100,
            base_speeds: vec![Speed::Walking(30)],
            gender: Gender::gen(&mut rng),
            height: 75,
            size: Size::Medium,
            weight: 300,
        };
        insta::assert_yaml_snapshot!(characteristics);
    }

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_characteristics_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let characteristics = CharacteristicDetails {
            age: 100,
            base_speeds: vec![Speed::Walking(30)],
            gender: Gender::gen(&mut rng),
            height: 75,
            size: Size::Medium,
            weight: 300,
        };
        insta::assert_snapshot!(format!("{}", characteristics));
    }
}
