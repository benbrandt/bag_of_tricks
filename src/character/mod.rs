mod ability;

use ability::{AbilityScoreIncreases, AbilityScores};
use rand::Rng;
use std::fmt;

/// Character information
#[derive(Debug)]
pub struct Character {
    abilities: AbilityScores,
}

impl Character {
    /// Generate a new random character
    pub fn new(rng: &mut impl Rng) -> Self {
        Self {
            abilities: AbilityScores::new(rng, AbilityScoreIncreases::default()),
        }
    }
}

impl fmt::Display for Character {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}", self.abilities)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_character_new() {
        let mut rng = Pcg64::from_entropy();
        let Character {
            abilities:
                AbilityScores {
                    charisma,
                    constitution,
                    dexterity,
                    intelligence,
                    strength,
                    wisdom,
                },
        } = Character::new(&mut rng);
        assert!(charisma.base >= 3 && charisma.base <= 18);
        assert!(constitution.base >= 3 && constitution.base <= 18);
        assert!(dexterity.base >= 3 && dexterity.base <= 18);
        assert!(intelligence.base >= 3 && intelligence.base <= 18);
        assert!(strength.base >= 3 && strength.base <= 18);
        assert!(wisdom.base >= 3 && wisdom.base <= 18);
    }
}
