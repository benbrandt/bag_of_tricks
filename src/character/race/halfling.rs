use rand::prelude::IteratorRandom;
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::fmt;
use strum::IntoEnumIterator;
use strum_macros::{Display, EnumIter};

use super::Race;
use crate::{
    character::{
        ability::{AbilityScore, AbilityScoreType, AbilityScores},
        characteristics::{AgeRange, Characteristics},
        features::Feature,
    },
    citation::{Book, Citation, Citations},
};

const AGE_RANGE: AgeRange = AgeRange(1..=150);
mod names {
    use rand::{prelude::IteratorRandom, Rng};

    use crate::character::characteristics::{Characteristics, Gender};

    const FAMILY: &[&str] = &[
        "Brushgather",
        "Goodbarrel",
        "Greenbottle",
        "High-hill",
        "Hilltopple",
        "Leagallow",
        "Tealeaf",
        "Thorngage",
        "Tosscobble",
        "Underbough",
    ];

    const FEMALE: &[&str] = &[
        "Andry",
        "Bree",
        "Callie",
        "Cora",
        "Euphemia",
        "Jillian",
        "Kithri",
        "Lavinia",
        "Lidda",
        "Merla",
        "Nedda",
        "Paela",
        "Portia",
        "Seraphina",
        "Shaena",
        "Trym",
        "Vani",
        "Verna",
    ];

    const MALE: &[&str] = &[
        "Alton", "Ander", "Cade", "Corrin", "Eldon", "Errich", "Finnan", "Garret", "Lindal",
        "Lyle", "Merric", "Milo", "Osborn", "Perrin", "Reed", "Roscoe", "Wellby",
    ];

    pub(crate) fn gen_name(
        rng: &mut impl Rng,
        Characteristics { gender, .. }: &Characteristics,
    ) -> String {
        let first_names = match gender {
            Gender::Female => FEMALE,
            Gender::Male => MALE,
        };
        format!(
            "{} {}",
            first_names.iter().choose(rng).unwrap(),
            FAMILY.iter().choose(rng).unwrap(),
        )
    }
}

#[derive(Debug, Deserialize, Display, EnumIter, PartialEq, Serialize)]
enum HalflingSubrace {
    Lightfoot,
    Stout,
}

#[derive(Deserialize, Serialize)]
pub(crate) struct Halfling {
    subrace: HalflingSubrace,
}

#[typetag::serde]
impl Race for Halfling {
    fn gen(rng: &mut impl Rng) -> (Box<dyn Race>, String, Characteristics) {
        let race = Box::new(Self {
            subrace: HalflingSubrace::iter().choose(rng).unwrap(),
        });
        let characteristics = Characteristics::gen(rng, &AGE_RANGE);
        let name = names::gen_name(rng, &characteristics);
        (race, name, characteristics)
    }

    fn abilities(&self) -> AbilityScores {
        AbilityScores(vec![
            AbilityScore(AbilityScoreType::Dexterity, 2),
            match self.subrace {
                HalflingSubrace::Lightfoot => AbilityScore(AbilityScoreType::Charisma, 1),
                HalflingSubrace::Stout => AbilityScore(AbilityScoreType::Constitution, 1),
            },
        ])
    }

    fn citations(&self) -> Citations {
        let race = Citation {
            book: Book::PHB,
            page: 26,
        };
        let subrace = match self.subrace {
            HalflingSubrace::Lightfoot | HalflingSubrace::Stout => Citation {
                book: Book::PHB,
                page: 28,
            },
        };
        Citations(vec![race, subrace])
    }

    fn features(&self) -> Vec<Feature> {
        let mut features = vec![Feature {
            title: "Ability Score Increase",
            description: "Your Dexterity score increases by 2.",
            citation: Citation {
                book: Book::PHB,
                page: 28,
            },
        }];
        features.extend(match self.subrace {
            HalflingSubrace::Lightfoot => vec![Feature {
                title: "Ability Score Increase",
                description: "Your Charisma score increases by 1.",
                citation: Citation {
                    book: Book::PHB,
                    page: 28,
                },
            }],
            HalflingSubrace::Stout => vec![Feature {
                title: "Ability Score Increase",
                description: "Your Constitution score increases by 2.",
                citation: Citation {
                    book: Book::PHB,
                    page: 28,
                },
            }],
        });
        features
    }
}

impl fmt::Display for Halfling {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{} Halfling", self.subrace)
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    #[test]
    fn test_snapshot() {
        let mut rng = Pcg64::seed_from_u64(1);
        let halfling = Halfling::gen(&mut rng);
        insta::assert_yaml_snapshot!(halfling);
    }

    #[test]
    fn test_snapshot_display() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_snapshot!(format!("{}", halfling));
        }
    }

    #[test]
    fn test_snapshot_abilities() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_yaml_snapshot!(halfling.abilities());
        }
    }

    #[test]
    fn test_snapshot_citations() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_yaml_snapshot!(halfling.citations());
        }
    }

    #[test]
    fn test_snapshot_features() {
        for subrace in HalflingSubrace::iter() {
            let halfling = Halfling { subrace };
            insta::assert_yaml_snapshot!(halfling.features());
        }
    }
}
