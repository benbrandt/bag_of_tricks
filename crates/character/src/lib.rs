#![deny(clippy::all)]
#![warn(clippy::pedantic)]

mod ability;
mod background;
mod equipment;
mod features;
mod proficiencies;
mod race;

use std::{fmt, writeln};

use ability::{AbilityScores, Skill};
use alignment::{Alignment, AlignmentInfluences, Attitude, Morality};
use attack::{DamageType, Resistances};
use background::{Background, BackgroundOption};
use backstory::Backstory;
use characteristics::{Appearance, CharacteristicDetails, Speed};
use equipment::{currency::Coin, Equipment, EquipmentOption};
use features::{Feature, Features};
use languages::Language;
use personality::Personality;
use proficiencies::Proficiency;
use race::{Race, RaceOptions};
use rand::{prelude::SliceRandom, Rng};
use serde::{Deserialize, Serialize};
use strum::IntoEnumIterator;
use trinkets::{TrinketOption, Trinkets};

/// Character information. Mostly stores random choices made for this character.
#[derive(Default, Deserialize, Serialize)]
pub struct Character {
    /// Ability scores for the character.
    abilities: AbilityScores,
    /// The character's alignment.
    alignment: Option<Alignment>,
    /// The character's background choice.
    background: Option<Box<dyn Background>>,
    /// Characteristics of the character.
    characteristics: Option<CharacteristicDetails>,
    /// Currency
    coins: (Coin, u8),
    /// Equipment randomly chosen for the character.
    equipment: Vec<Equipment>,
    /// Languages randomly chosen for the character.
    languages: Vec<Language>,
    /// Current level of the character.
    level: u8,
    /// Character's name.
    name: String,
    /// Personality traits of the chracacter.
    personality: Option<Personality>,
    /// Proficiencies for the character.
    proficiencies: Vec<Proficiency>,
    /// Race randomly chosen for the character.
    race: Option<Box<dyn Race>>,
}

impl Character {
    /// Generate a new random character
    ///
    /// The methodolgy is to gather together as many static inputs as possible (based on some initial random choices),
    /// and then have those feed into later choices.
    ///
    /// Steps are as follows:
    /// 1. Randomly choose a Race (which also generates a name and some physical characteristics)
    /// 2. Roll ability scores (and apply the racial ability increases)
    /// 3. Randomly choose a background and personality traits (weighted by highest skill modifiers)
    /// 4. Choose alignment (weighted based on inputs from race and background)
    /// 5. Randomly choose any additional languages
    /// 6. Randomly choose proficiencies, weighted towards optimal ones based on what is known about the character so far
    pub fn gen(rng: &mut impl Rng) -> Self {
        let mut character = Self {
            level: 1,
            ..Self::default()
        };
        let (race, name, characteristics) = RaceOptions::gen(rng);
        let mut abilities = AbilityScores::gen(rng);
        abilities.increase(race.abilities());
        character.abilities = abilities;
        let background = BackgroundOption::gen(
            rng,
            &character.abilities,
            &character.proficiencies,
            character.proficiency_bonus(),
        );
        character.race = Some(race);
        character.name = name;
        character.characteristics = Some(characteristics);
        character.background = Some(background);
        character.gen_personality(rng);
        character.gen_alignment(rng);
        character.gen_languages(rng);
        character.gen_proficiences(rng);
        character.gen_equipment(rng);
        character
    }

    /// Generate character's alignment, feeding in any inputs we have for attitude and morality.
    fn gen_alignment(&mut self, rng: &mut impl Rng) {
        self.alignment = Some(Alignment::gen(rng, &self.attitude(), &self.morality()));
    }

    /// Generate any additional equipment.
    fn gen_equipment(&mut self, rng: &mut impl Rng) {
        // Choose a trinket
        let mut addl_equipment = vec![EquipmentOption::Trinket(None, None, true)];

        if let Some(background) = self.background.as_ref() {
            self.coins = background.coins();
            self.equipment.extend(background.equipment());
            addl_equipment.extend(background.addl_equipment());
        }

        addl_equipment.sort();
        for option in addl_equipment {
            self.equipment.push(option.gen(
                rng,
                &self.equipment,
                &self.proficiencies,
                &self.trinket_options(),
            ));
        }
    }

    /// Generate any additional languages, ensuring no overlap with current languages.
    fn gen_languages(&mut self, rng: &mut impl Rng) {
        let mut languages = vec![];

        if let Some(race) = self.race.as_ref() {
            languages.extend(race.languages());
        }
        if let Some(background) = self.background.as_ref() {
            languages.extend(background.languages());
        }

        let mut addl_languages = vec![];
        if let Some(race) = self.race.as_ref() {
            addl_languages.push(race.addl_languages());
        }
        if let Some(background) = self.background.as_ref() {
            addl_languages.push(background.addl_languages());
        }

        // Handle any dupes across these options
        for l in languages {
            if self.languages.contains(&l) {
                self.languages
                    .extend(Language::gen(rng, &self.languages, (1, None)));
            } else {
                self.languages.push(l);
            }
        }

        for a in addl_languages {
            self.languages
                .extend(Language::gen(rng, &self.languages, a));
        }
    }

    /// Generate personality descriptions from the associated constants
    fn gen_personality(&mut self, rng: &mut impl Rng) {
        let mut bonds = vec![];
        let mut flaws = vec![];
        let mut ideals = vec![];
        let mut traits = vec![];
        if let Some(race) = self.race.as_ref() {
            bonds.extend(race.bonds());
            flaws.extend(race.flaws());
            ideals.extend(race.ideals());
            traits.extend(race.traits());
        }
        if let Some(background) = self.background.as_ref() {
            bonds.extend(background.bonds());
            flaws.extend(background.flaws());
            ideals.extend(background.ideals());
            traits.extend(background.traits());
        }
        self.personality = Some(Personality {
            bond: bonds.choose(rng).unwrap().clone(),
            flaw: flaws.choose(rng).unwrap().clone(),
            ideal: ideals.choose(rng).unwrap().clone(),
            traits: traits.choose_multiple(rng, 2).cloned().collect(),
        });
    }

    /// Generate additional proficiencies, each looking at the current character sheet.
    ///
    /// They are generated from the most limited sets of options to greatest, to avoid overlapping choices.
    fn gen_proficiences(&mut self, rng: &mut impl Rng) {
        // Static choices
        let mut proficiencies = vec![];
        let mut addl_proficiencies = vec![];
        if let Some(race) = self.race.as_ref() {
            proficiencies.extend(race.proficiencies());
            addl_proficiencies.extend(race.addl_proficiencies());
        }
        if let Some(background) = self.background.as_ref() {
            proficiencies.extend(background.proficiencies());
            addl_proficiencies.extend(background.addl_proficiencies());
        }
        // Handle any dupes across these options
        for p in proficiencies {
            if self.proficiencies.contains(&p) {
                self.proficiencies.extend(p.gen_replacement(
                    rng,
                    &self.abilities,
                    &self.proficiencies,
                    self.proficiency_bonus(),
                ));
            } else {
                self.proficiencies.push(p);
            }
        }

        // Sort so that the options with the least amount are chosen first.
        addl_proficiencies.sort();
        for option in addl_proficiencies {
            self.proficiencies.extend(option.gen(
                rng,
                &self.abilities,
                &self.proficiencies,
                self.proficiency_bonus(),
            ));
        }
    }

    /// Return the character's proficiency bonus based on their level.
    fn proficiency_bonus(&self) -> i16 {
        match self.level {
            0..=4 => 2,
            5..=8 => 3,
            9..=12 => 4,
            13..=16 => 5,
            17..=u8::MAX => 6,
        }
    }

    /// Return the speeds of the character.
    fn speeds(&self) -> Vec<Speed> {
        self.characteristics
            .as_ref()
            .map(|c| c.base_speeds.clone())
            .unwrap_or_default()
    }
}

/// Combine all attitude and morality influences for the character (race and personality)
impl AlignmentInfluences for Character {
    fn attitude(&self) -> Vec<Attitude> {
        let mut attitude = vec![];
        if let Some(race) = self.race.as_ref() {
            attitude.extend(race.attitude());
        }
        if let Some(personality) = self.personality.as_ref() {
            attitude.extend(personality.attitude());
        }
        attitude
    }

    fn morality(&self) -> Vec<Morality> {
        let mut morality = vec![];
        if let Some(race) = self.race.as_ref() {
            morality.extend(race.morality());
        }
        if let Some(personality) = self.personality.as_ref() {
            morality.extend(personality.morality());
        }
        morality
    }
}

impl Appearance for Character {
    fn appearance(&self) -> Vec<String> {
        let mut appearance = vec![];
        if let Some(race) = self.race.as_ref() {
            appearance.extend(race.appearance());
        }
        appearance
    }
}

/// Combine all backstory items for the character.
impl Backstory for Character {
    fn backstory(&self) -> Vec<String> {
        let mut backstory = vec![];
        if let Some(race) = self.race.as_ref() {
            backstory.extend(race.backstory());
        }
        if let Some(background) = self.background.as_ref() {
            backstory.extend(background.backstory());
        }
        backstory
    }
}

/// Combine all features and traits for the characters (race and background)
impl Features for Character {
    fn features(&self) -> Vec<Feature> {
        let mut features = vec![];
        if let Some(race) = self.race.as_ref() {
            features.extend(race.features());
        }
        if let Some(background) = self.background.as_ref() {
            features.extend(background.features());
        }
        features
    }
}

/// Combine all resistances the character has.
impl Resistances for Character {
    fn immunities(&self) -> Vec<DamageType> {
        self.race
            .as_ref()
            .map(|r| r.immunities())
            .unwrap_or_default()
    }

    fn resistances(&self) -> Vec<DamageType> {
        self.race
            .as_ref()
            .map(|r| r.resistances())
            .unwrap_or_default()
    }
}

impl Trinkets for Character {
    fn trinket_options(&self) -> Vec<TrinketOption> {
        let mut options = vec![TrinketOption::Default];
        if let Some(race) = self.race.as_ref() {
            options.extend(race.trinket_options());
        }
        options.sort();
        options.dedup();
        options
    }
}

/// Render a text version of the character. Useful for CLI or other output.
impl fmt::Display for Character {
    #[allow(clippy::too_many_lines)]
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "CHARACTER NAME: {}", self.name)?;
        if let Some(race) = self.race.as_ref() {
            writeln!(f, "RACE: {} ({})", race, race.citations())?;
        }
        if let Some(background) = self.background.as_ref() {
            writeln!(f, "BACKGROUND: {} ({})", background, background.citations())?;
        }
        writeln!(f, "ALIGNMENT: {}", self.alignment.as_ref().unwrap())?;
        writeln!(f, "LEVEL: {}", self.level)?;
        writeln!(f)?;
        writeln!(f, "{}", self.abilities)?;
        writeln!(f, "SKILLS:")?;
        writeln!(f, "PROF  MOD  SKILL            BONUS:")?;
        for skill in Skill::iter() {
            writeln!(
                f,
                "{:4}  {}  {:15}  {:+}",
                if skill.proficient(&self.proficiencies,) {
                    " X"
                } else {
                    ""
                },
                skill.ability_score_type(),
                skill,
                skill.modifier(
                    &self.abilities,
                    &self.proficiencies,
                    self.proficiency_bonus(),
                ),
            )?;
        }
        writeln!(f)?;
        writeln!(f, "SPEED:")?;
        for speed in self.speeds() {
            writeln!(f, "{}", speed)?;
        }
        writeln!(f, "PROFICIENCY BONUS: {:+}", self.proficiency_bonus())?;
        writeln!(
            f,
            "RESISTANCES: {}",
            self.resistances()
                .iter()
                .map(|r| format!("{}", r))
                .collect::<Vec<String>>()
                .join(", ")
        )?;
        writeln!(
            f,
            "IMMUNITIES: {}",
            self.immunities()
                .iter()
                .map(|r| format!("{}", r))
                .collect::<Vec<String>>()
                .join(", ")
        )?;
        writeln!(
            f,
            "LANGUAGES: {}",
            self.languages
                .iter()
                .map(|r| format!("{}", r))
                .collect::<Vec<String>>()
                .join(", ")
        )?;
        writeln!(
            f,
            "PROFICIENCIES: {}",
            self.proficiencies
                .iter()
                .filter_map(|p| match p {
                    Proficiency::Skill(_) => None,
                    Proficiency::Armor(_)
                    | Proficiency::Tool(_)
                    | Proficiency::Weapon(_)
                    | Proficiency::Vehicle(_) => Some(format!("{:?}", p)),
                })
                .collect::<Vec<String>>()
                .join(", ")
        )?;
        writeln!(f)?;
        writeln!(f, "CHARACTERISTICS:")?;
        if let Some(characteristics) = self.characteristics.as_ref() {
            writeln!(f, "{}", characteristics)?;
        }
        if let Some(personality) = self.personality.as_ref() {
            writeln!(f, "{}", personality)?;
        }
        writeln!(f, "EQUIPMENT")?;
        writeln!(
            f,
            "{}",
            self.equipment
                .iter()
                .map(|e| format!("{:?}", e))
                .collect::<Vec<String>>()
                .join(", ")
        )?;
        writeln!(f, "COINS: {}{}", self.coins.1, self.coins.0)?;
        writeln!(f)?;
        writeln!(f, "FEATURES AND TRAITS:")?;
        for feature in self.features() {
            writeln!(f, "- {}", feature)?;
        }
        writeln!(f)?;
        writeln!(f, "APPEARANCE:")?;
        for appearance in self.appearance() {
            writeln!(f, "{}", appearance)?;
        }
        writeln!(f)?;
        writeln!(f, "BACKSTORY:")?;
        for backstory in self.backstory() {
            writeln!(f, "{}", backstory)?;
        }
        write!(f, "")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_character() {
        let mut rng = Pcg64::seed_from_u64(1);
        let character = Character::gen(&mut rng);
        insta::assert_yaml_snapshot!(character);
    }

    /// Verify that our snapshot remains the same.
    #[test]
    // #[ignore = "linux vs mac differences"]
    fn test_character_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let character = Character::gen(&mut rng);
        insta::assert_display_snapshot!(character);
    }
}
