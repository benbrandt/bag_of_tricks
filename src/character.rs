mod ability;
mod characteristics;
mod features;
mod race;

use features::Feature;
use rand::Rng;
use serde::{Deserialize, Serialize};
use std::fmt;

use ability::AbilityScores;
use characteristics::Characteristics;
use race::{Race, RaceOptions};

/// Character information
#[derive(Deserialize, Serialize)]
pub struct Character {
    abilities: AbilityScores,
    characteristics: Characteristics,
    name: String,
    race: Box<dyn Race>,
}

impl Character {
    /// Generate a new random character
    pub fn new(rng: &mut impl Rng) -> Self {
        let (race, name, characteristics) = RaceOptions::gen(rng);
        let mut abilities = AbilityScores::new(rng);
        abilities.extend(race.abilities());
        Self {
            abilities,
            characteristics,
            name,
            race,
        }
    }

    fn features(&self) -> Vec<Feature> {
        self.race.features()
    }
}

impl fmt::Display for Character {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        writeln!(f, "CHARACTER NAME: {}", self.name)?;
        writeln!(f, "RACE: {} ({})", self.race, self.race.citations())?;
        writeln!(f, "")?;
        writeln!(f, "{}", self.abilities)?;
        writeln!(f, "CHARACTERISTICS:")?;
        writeln!(f, "{}", self.characteristics)?;
        writeln!(f, "FEATURES AND TRAITS:")?;
        for feature in self.features() {
            writeln!(f, "{}", feature)?;
        }
        write!(f, "")
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use rand::SeedableRng;
    use rand_pcg::Pcg64;

    /// Verify that our snapshot remains the same.
    #[test]
    fn test_character_display() {
        let mut rng = Pcg64::seed_from_u64(1);
        let character = Character::new(&mut rng);
        // Struct Snapshot
        insta::assert_yaml_snapshot!(character);
        // fmt::Display Snapshot
        insta::assert_snapshot!(format!("{}", character));
    }
}
