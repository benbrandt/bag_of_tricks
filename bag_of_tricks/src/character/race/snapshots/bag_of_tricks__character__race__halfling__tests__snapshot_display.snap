---
source: src/character/race/halfling.rs
expression: "HalflingSubrace::iter().map(|subrace|\n                                format!(\"{}\", Halfling\n                                        {\n                                            subrace\n                                        })).collect::<Vec<String>>().join(\"\\n\\n\")"

---
Ghostwise Halfling

Lightfoot Halfling

Stout Halfling
